enhancements/20220531-storage-tiering.md

```diff
@@ -0,0 +1,142 @@
+  - For the above two points, save source provisioner, storageClassName,ImageID in snapshot's annotations during volume snapshot.
+
+##### 3.2 Show Volume
+```
+if provisioner is driver.longhorn.io
+  show diskSelector and nodeSelector 
+else
+  show storageClassName
+fi
+```
+
++ `Volumes` > volume
+![image](./20220531-storage-tiering/show-volume.png)
+
+#### 4 Image
+Since the storageClass of backing image is created dynamically, we only need to let users select diskSelector and nodeSelector during vm image creating.
+
+##### 4.1 Create Image
+
+Select diskSelector,nodeSelector (default to `any available`)
++ `Images` > `Create`
+  ![image](./20220531-storage-tiering/create-image.png)
++ `Volumes` > volume > `Export`
+  ![image](./20220531-storage-tiering/volume-export.png)
+
+##### 4.2 Show Image
+Show diskSelector,nodeSelector, if it exits
+
++ `Images` > image
+  ![image](./20220531-storage-tiering/show-image.png)
+
+
+### API changes
+
+Add a new field: `extraStorageClassParameters`, type `map[string]string`, to `harvesterhci.io.virtualmachineimages.spec`.
+
+If `extraStorageClassParameters` are allowed to be modified, the backend deletion and reconstruction of the storageClass being used may cause unpredictable consequences.
+And it is difficult to know from the storageClass information what the parameters were when the PVC was created.
+This makes displaying diskSelector and nodeSelector on the UI potentially inaccurate.
+Currently, the URL of image is also not allowed to be modified by the user on the UI
+So users should not be allowed to modify the `extraStorageClassParameters` contents.
+For different scheduling rules for the same image file, users need to create different images.
+
+```yaml
+spec:
+  checksum: ""
+  displayName: ubuntu-20.04-minimal-cloudimg-amd64.img
+  extraStorageClassParameters:
+    nodeSelector: "large"
+    diskSelector: "hdd"
+    migratable: "true"
+    numberOfReplicas: "3"
+    staleReplicaTimeout: "30"
+```
+
+## Design
+
+### Implementation Overview
+
+1. Specify diskSelector and nodeSelector in `spec.extraStorageClassParameters` when vm image creating or updating.
+2. The `vm image mutator` in `harvester-webhook` will autofill other extraStorageClassParameters according to the `image-storage-class-parameters` settings.
+3. The `vm image controller` in `harvester` will create or update a backing image storageClass according to the `spec.extraStorageClassParameters`.
+
+#### New Settings `image-default-storage-class-parameters`
+
+Default to `{"numberOfReplicas":"3","staleReplicaTimeout":"30","migratable":"true"}`
+```go
+ImageDefaultStorageClassParameters   = NewSetting("image-default-storage-class-parameters", `{"numberOfReplicas":"3","staleReplicaTimeout":"30","migratable":"true"}`)
+```
+
+#### Add vm image mutator to harvester-webhook
+
+![image](./20220531-storage-tiering/backingimage.png)
```

Comment:
1. let's remove the other SC for now since they are out of the scope.
2. In the architecture flow, the Harvester will list/watch the `vmimage`.

Reply:
updated